File: python/mighty_sdk/depth.py
import math
from typing import Any, Callable, Dict, Optional

import mighty_protocol as mp


def _timestamp_ns(frame: Dict[str, Any]) -> int:
    try:
        return int(frame.get("timestamp_ns", 0) or 0)
    except (TypeError, ValueError):
        return 0


def _canonical_channel(channel: str) -> str:
    value = str(channel or "").strip().lower()
    if value in ("preview", "left"):
        return "cam0"
    if value == "right":
        return "cam1"
    return value


def depth_at_meters(frame: Dict[str, Any], x: int, y: int) -> Optional[float]:
    width = int(frame.get("width", 0) or 0)
    height = int(frame.get("height", 0) or 0)
    px = int(x)
    py = int(y)
    values = frame.get("depth_mm")
    if values is None or px < 0 or py < 0 or px >= width or py >= height:
        return None
    sample = int(values[py * width + px])
    if sample == int(frame.get("invalid_value", 0)):
        return None
    scale = float(frame.get("depth_scale_m", 0.001))
    return sample * scale if math.isfinite(scale) and scale > 0.0 else None


def _raw_to_rgba(image: Dict[str, Any]):
    width = int(image.get("width", 0) or 0)
    height = int(image.get("height", 0) or 0)
    data = bytes(image.get("data", b""))
    fmt = int(image.get("format", 0) or 0)
    pixels = width * height
    if width <= 0 or height <= 0:
        return None
    rgba = bytearray(pixels * 4)
    if fmt == mp.RAW_FORMAT["GRAY8"] or fmt in (mp.RAW_FORMAT["YUV420SP"], mp.RAW_FORMAT["YUV420P"]):
        if len(data) < pixels:
            return None
        for i in range(pixels):
            value = data[i]
            rgba[i * 4:i * 4 + 4] = bytes((value, value, value, 255))
    elif fmt in (mp.RAW_FORMAT["RGB24"], mp.RAW_FORMAT["BGR24"]):
        if len(data) < pixels * 3:
            return None
        bgr = fmt == mp.RAW_FORMAT["BGR24"]
        for i in range(pixels):
            source = i * 3
            red, green, blue = data[source:source + 3]
            if bgr:
                red, blue = blue, red
            rgba[i * 4:i * 4 + 4] = bytes((red, green, blue, 255))
    elif fmt in (mp.RAW_FORMAT["RGBA32"], mp.RAW_FORMAT["BGRA32"]):
        if len(data) < pixels * 4:
            return None
        bgra = fmt == mp.RAW_FORMAT["BGRA32"]
        for i in range(pixels):
            source = i * 4
            red, green, blue, alpha = data[source:source + 4]
            if bgra:
                red, blue = blue, red
            rgba[i * 4:i * 4 + 4] = bytes((red, green, blue, alpha))
    else:
        return None
    return {"width": width, "height": height, "rgba": rgba}


def _distort_normalized(x: float, y: float, frame: Dict[str, Any]):
    coeffs = list(frame.get("distortion", (0.0, 0.0, 0.0, 0.0)))[:4]
    coeffs += [0.0] * (4 - len(coeffs))
    camera_model = int(frame.get("camera_model", mp.DEPTH_CAMERA_MODEL["PINHOLE"]))
    distortion_model = int(frame.get("distortion_model", mp.DEPTH_DISTORTION_MODEL["NONE"]))
    if camera_model == mp.DEPTH_CAMERA_MODEL["DOUBLE_SPHERE"]:
        xi, alpha = coeffs[:2]
        radius2 = x * x + y * y
        if radius2 < 1e-16:
            return x, y
        d1 = math.sqrt(radius2 + 1.0)
        shifted = xi * d1 + 1.0
        d2 = math.sqrt(radius2 + shifted * shifted)
        denominator = alpha * d2 + (1.0 - alpha) * shifted
        return (x / denominator, y / denominator) if abs(denominator) > 1e-12 else (math.nan, math.nan)
    if distortion_model == mp.DEPTH_DISTORTION_MODEL["NONE"]:
        return x, y
    if distortion_model == mp.DEPTH_DISTORTION_MODEL["EQUIDISTANT"]:
        k1, k2, k3, k4 = coeffs
        radius = math.hypot(x, y)
        if radius < 1e-8:
            return x, y
        theta = math.atan(radius)
        theta2 = theta * theta
        theta_d = theta * (1.0 + k1 * theta2 + k2 * theta2 ** 2 + k3 * theta2 ** 3 + k4 * theta2 ** 4)
        scale = theta_d / radius
        return x * scale, y * scale
    k1, k2, p1, p2 = coeffs
    x2, y2 = x * x, y * y
    xy = x * y
    radius2 = x2 + y2
    radial = 1.0 + k1 * radius2 + k2 * radius2 * radius2
    return (
        x * radial + 2.0 * p1 * xy + p2 * (radius2 + 2.0 * x2),
        y * radial + p1 * (radius2 + 2.0 * y2) + 2.0 * p2 * xy,
    )


def rectify_image_to_depth(image: Dict[str, Any], depth: Dict[str, Any], require_matching_timestamp: bool = True):
    image_ts = _timestamp_ns(image)
    depth_ts = _timestamp_ns(depth)
    if require_matching_timestamp and image_ts > 0 and depth_ts > 0 and image_ts != depth_ts:
        raise ValueError("source image and depth timestamps do not match")
    decoded = _raw_to_rgba(image)
    if decoded is None:
        raise ValueError("source image format cannot be decoded to RGBA")
    width = int(depth.get("width", 0) or 0)
    height = int(depth.get("height", 0) or 0)
    depth_k = tuple(float(v) for v in depth.get("depth_intrinsics", ()))
    source_k = tuple(float(v) for v in depth.get("source_intrinsics", ()))
    if width <= 0 or height <= 0 or len(depth_k) != 4 or len(source_k) != 4 or \
            depth_k[0] <= 0.0 or depth_k[1] <= 0.0 or source_k[0] <= 0.0 or source_k[1] <= 0.0:
        raise ValueError("depth frame has invalid camera geometry")
    calibration_width = int(depth.get("source_width", decoded["width"]) or decoded["width"])
    calibration_height = int(depth.get("source_height", decoded["height"]) or decoded["height"])
    scale_x = decoded["width"] / calibration_width
    scale_y = decoded["height"] / calibration_height
    sfx, sfy = source_k[0] * scale_x, source_k[1] * scale_y
    scx, scy = source_k[2] * scale_x, source_k[3] * scale_y
    source_rgba = decoded["rgba"]
    output = bytearray(width * height * 4)
    for py in range(height):
        for px in range(width):
            ray_x = (px - depth_k[2]) / depth_k[0]
            ray_y = (py - depth_k[3]) / depth_k[1]
            distorted_x, distorted_y = _distort_normalized(ray_x, ray_y, depth)
            sample_x = sfx * distorted_x + scx
            sample_y = sfy * distorted_y + scy
            destination = (py * width + px) * 4
            if not math.isfinite(sample_x) or not math.isfinite(sample_y) or \
                    sample_x < 0.0 or sample_y < 0.0 or \
                    sample_x > decoded["width"] - 1 or sample_y > decoded["height"] - 1:
                output[destination + 3] = 255
                continue
            x0, y0 = int(math.floor(sample_x)), int(math.floor(sample_y))
            x1, y1 = min(decoded["width"] - 1, x0 + 1), min(decoded["height"] - 1, y0 + 1)
            ax, ay = sample_x - x0, sample_y - y0
            indexes = [
                (y0 * decoded["width"] + x0) * 4,
                (y0 * decoded["width"] + x1) * 4,
                (y1 * decoded["width"] + x0) * 4,
                (y1 * decoded["width"] + x1) * 4,
            ]
            for channel in range(4):
                top = source_rgba[indexes[0] + channel] + (source_rgba[indexes[1] + channel] - source_rgba[indexes[0] + channel]) * ax
                bottom = source_rgba[indexes[2] + channel] + (source_rgba[indexes[3] + channel] - source_rgba[indexes[2] + channel]) * ax
                output[destination + channel] = round(top + (bottom - top) * ay)
    return {
        "width": width,
        "height": height,
        "rgba": output,
        "timestamp_ns": depth_ts,
        "frame_id": depth.get("frame_id", "cam0_rectified"),
    }


class RgbdSynchronizer:
    def __init__(self, on_pair: Callable[[Dict[str, Any]], None], max_entries: int = 64):
        if not callable(on_pair):
            raise ValueError("RgbdSynchronizer requires an on_pair callback")
        self.on_pair = on_pair
        self.max_entries = max(2, int(max_entries))
        self.images: Dict[str, Dict[str, Any]] = {}
        self.depth: Dict[str, Dict[str, Any]] = {}

    def _key(self, channel: str, timestamp_ns: int) -> str:
        return f"{_canonical_channel(channel)}:{int(timestamp_ns)}"

    def _trim(self, cache: Dict[str, Dict[str, Any]]) -> None:
        while len(cache) > self.max_entries:
            cache.pop(next(iter(cache)))

    def push_image(self, image: Dict[str, Any]) -> None:
        if image and image.get("kind") == "stereo_raw":
            self.push_image(image.get("left"))
            self.push_image(image.get("right"))
            return
        if not image:
            return
        timestamp_ns = _timestamp_ns(image)
        if timestamp_ns <= 0:
            return
        key = self._key(image.get("channel_alias") or image.get("channel", ""), timestamp_ns)
        depth = self.depth.pop(key, None)
        if depth is not None:
            self.on_pair({"image": image, "depth": depth})
            return
        self.images[key] = image
        self._trim(self.images)

    def push_depth(self, depth: Dict[str, Any]) -> None:
        if not depth:
            return
        timestamp_ns = _timestamp_ns(depth)
        if timestamp_ns <= 0:
            return
        key = self._key(depth.get("source_channel", ""), timestamp_ns)
        image = self.images.pop(key, None)
        if image is not None:
            self.on_pair({"image": image, "depth": depth})
            return
        self.depth[key] = depth
        self._trim(self.depth)

    def clear(self) -> None:
        self.images.clear()
        self.depth.clear()

File: js/sdk/depth.js
import {
  DEPTH_CAMERA_MODEL,
  DEPTH_DISTORTION_MODEL,
  DEPTH_ENCODING,
} from "../core/protocol.js";
import { decodeRawToRgb } from "./image.js";

function canonicalChannel(channel) {
  const value = String(channel || "").trim().toLowerCase();
  if (value === "preview" || value === "left") return "cam0";
  if (value === "right") return "cam1";
  return value;
}

function frameTimestamp(frame) {
  const value = frame?.timestampNs ?? frame?.timestamp_ns ?? 0n;
  try {
    return BigInt(value || 0);
  } catch (_) {
    return 0n;
  }
}

function rgbdKey(channel, timestampNs) {
  return `${canonicalChannel(channel)}:${BigInt(timestampNs || 0).toString()}`;
}

export function depthAtMeters(frame, x, y) {
  const width = Number(frame?.width || 0);
  const height = Number(frame?.height || 0);
  const px = Math.floor(Number(x));
  const py = Math.floor(Number(y));
  const values = frame?.depthMm;
  if (!values || width <= 0 || height <= 0 ||
      px < 0 || py < 0 || px >= width || py >= height) return null;
  const sample = Number(values[py * width + px]);
  const invalid = Number(frame?.invalidValue ?? 0);
  const scale = Number(frame?.depthScaleM ?? 0.001);
  if (!Number.isFinite(sample) || sample === invalid || !Number.isFinite(scale) || scale <= 0) return null;
  return sample * scale;
}

// Google's compact Turbo approximation. The protocol stores metric values;
// this palette is presentation-only and can be replaced by applications.
function turboRgb(value) {
  const x = Math.max(0, Math.min(1, value));
  const k = [1, x, x * x, x * x * x, x ** 4, x ** 5];
  const dot = (coeffs) => coeffs.reduce((sum, c, i) => sum + c * k[i], 0);
  const clampByte = (v) => Math.max(0, Math.min(255, Math.round(v * 255)));
  return [
    clampByte(dot([0.13572138, 4.61539260, -42.66032258, 132.13108234, -152.94239396, 59.28637943])),
    clampByte(dot([0.09140261, 2.19418839, 4.84296658, -14.18503333, 4.27729857, 2.82956604])),
    clampByte(dot([0.10667330, 12.64194608, -60.58204836, 110.36276771, -89.90310912, 27.34824973])),
  ];
}

export function colorizeDepth(frame, { nearM = 0, farM = 10 } = {}) {
  const width = Number(frame?.width || 0);
  const height = Number(frame?.height || 0);
  const values = frame?.depthMm;
  if (!values || width <= 0 || height <= 0 || values.length < width * height) return null;
  const near = Number(nearM);
  const far = Number(farM);
  const span = far - near;
  if (!Number.isFinite(span) || span <= 0) return null;
  const invalid = Number(frame?.invalidValue ?? 0);
  const scale = Number(frame?.depthScaleM ?? 0.001);
  const rgba = new Uint8ClampedArray(width * height * 4);
  for (let i = 0; i < width * height; i += 1) {
    const sample = Number(values[i]);
    const out = i * 4;
    if (sample === invalid) {
      rgba[out + 3] = 255;
      continue;
    }
    const depthM = sample * scale;
    const [r, g, b] = turboRgb((far - depthM) / span);
    rgba[out] = r;
    rgba[out + 1] = g;
    rgba[out + 2] = b;
    rgba[out + 3] = 255;
  }
  return { width, height, rgba };
}

function distortNormalized(x, y, frame) {
  const coeffs = Array.from(frame?.distortion || [0, 0, 0, 0], Number);
  while (coeffs.length < 4) coeffs.push(0);
  const cameraModel = Number(frame?.cameraModel ?? DEPTH_CAMERA_MODEL.PINHOLE);
  const distortionModel = Number(frame?.distortionModel ?? DEPTH_DISTORTION_MODEL.NONE);
  if (cameraModel === DEPTH_CAMERA_MODEL.DOUBLE_SPHERE) {
    const [xi, alpha] = coeffs;
    const r2 = x * x + y * y;
    if (r2 < 1e-16) return [x, y];
    const d1 = Math.sqrt(r2 + 1);
    const shifted = xi * d1 + 1;
    const d2 = Math.sqrt(r2 + shifted * shifted);
    const denominator = alpha * d2 + (1 - alpha) * shifted;
    return Number.isFinite(denominator) && Math.abs(denominator) > 1e-12
      ? [x / denominator, y / denominator]
      : [Number.NaN, Number.NaN];
  }
  if (distortionModel === DEPTH_DISTORTION_MODEL.NONE) return [x, y];
  if (distortionModel === DEPTH_DISTORTION_MODEL.EQUIDISTANT) {
    const [k1, k2, k3, k4] = coeffs;
    const r = Math.hypot(x, y);
    if (r < 1e-8) return [x, y];
    const theta = Math.atan(r);
    const t2 = theta * theta;
    const thetaD = theta * (1 + k1 * t2 + k2 * t2 ** 2 + k3 * t2 ** 3 + k4 * t2 ** 4);
    const scale = thetaD / r;
    return [x * scale, y * scale];
  }
  const [k1, k2, p1, p2] = coeffs;
  const x2 = x * x;
  const y2 = y * y;
  const xy = x * y;
  const r2 = x2 + y2;
  const radial = 1 + k1 * r2 + k2 * r2 * r2;
  return [
    x * radial + 2 * p1 * xy + p2 * (r2 + 2 * x2),
    y * radial + p1 * (r2 + 2 * y2) + 2 * p2 * xy,
  ];
}

function sampleBilinearRgba(rgba, width, height, x, y, out, offset) {
  if (!Number.isFinite(x) || !Number.isFinite(y) || x < 0 || y < 0 || x > width - 1 || y > height - 1) {
    out[offset + 3] = 255;
    return;
  }
  const x0 = Math.floor(x);
  const y0 = Math.floor(y);
  const x1 = Math.min(width - 1, x0 + 1);
  const y1 = Math.min(height - 1, y0 + 1);
  const ax = x - x0;
  const ay = y - y0;
  const i00 = (y0 * width + x0) * 4;
  const i01 = (y0 * width + x1) * 4;
  const i10 = (y1 * width + x0) * 4;
  const i11 = (y1 * width + x1) * 4;
  for (let c = 0; c < 4; c += 1) {
    const top = rgba[i00 + c] + (rgba[i01 + c] - rgba[i00 + c]) * ax;
    const bottom = rgba[i10 + c] + (rgba[i11 + c] - rgba[i10 + c]) * ax;
    out[offset + c] = Math.round(top + (bottom - top) * ay);
  }
}

export function rectifyImageToDepth(imageFrame, depthFrame, { requireMatchingTimestamp = true } = {}) {
  if (!imageFrame || !depthFrame) return null;
  const imageTs = frameTimestamp(imageFrame);
  const depthTs = frameTimestamp(depthFrame);
  if (requireMatchingTimestamp && imageTs > 0n && depthTs > 0n && imageTs !== depthTs) {
    throw new Error("source image and depth timestamps do not match");
  }
  const decoded = imageFrame.rgba
    ? { width: Number(imageFrame.width), height: Number(imageFrame.height), rgba: imageFrame.rgba }
    : decodeRawToRgb(imageFrame);
  if (!decoded) throw new Error("source image format cannot be decoded to RGBA");
  const width = Number(depthFrame.width || 0);
  const height = Number(depthFrame.height || 0);
  const [dfx, dfy, dcx, dcy] = Array.from(depthFrame.depthIntrinsics || [], Number);
  const sourceK = Array.from(depthFrame.sourceIntrinsics || [], Number);
  if (width <= 0 || height <= 0 || ![dfx, dfy, dcx, dcy, ...sourceK].every(Number.isFinite) ||
      dfx <= 0 || dfy <= 0 || sourceK[0] <= 0 || sourceK[1] <= 0) {
    throw new Error("depth frame has invalid camera geometry");
  }
  const calibrationWidth = Number(depthFrame.sourceWidth || decoded.width);
  const calibrationHeight = Number(depthFrame.sourceHeight || decoded.height);
  const sx = decoded.width / calibrationWidth;
  const sy = decoded.height / calibrationHeight;
  const sfx = sourceK[0] * sx;
  const sfy = sourceK[1] * sy;
  const scx = sourceK[2] * sx;
  const scy = sourceK[3] * sy;
  const rgba = new Uint8ClampedArray(width * height * 4);
  for (let y = 0; y < height; y += 1) {
    for (let x = 0; x < width; x += 1) {
      const rayX = (x - dcx) / dfx;
      const rayY = (y - dcy) / dfy;
      const [distortedX, distortedY] = distortNormalized(rayX, rayY, depthFrame);
      sampleBilinearRgba(
        decoded.rgba, decoded.width, decoded.height,
        sfx * distortedX + scx, sfy * distortedY + scy,
        rgba, (y * width + x) * 4,
      );
    }
  }
  return {
    width,
    height,
    rgba,
    timestampNs: depthTs,
    frameId: depthFrame.frameId || "cam0_rectified",
  };
}

export class RgbdSynchronizer {
  constructor(onPair, { maxEntries = 64 } = {}) {
    if (typeof onPair !== "function") throw new Error("RgbdSynchronizer requires an onPair callback");
    this.onPair = onPair;
    this.maxEntries = Math.max(2, Number(maxEntries) || 64);
    this.images = new Map();
    this.depth = new Map();
  }

  _trim(cache) {
    while (cache.size > this.maxEntries) cache.delete(cache.keys().next().value);
  }

  pushImage(image) {
    if (image?.kind === "stereo_raw") {
      this.pushImage(image.left);
      this.pushImage(image.right);
      return;
    }
    const timestampNs = frameTimestamp(image);
    if (!image || timestampNs <= 0n) return;
    const key = rgbdKey(image.channelAlias || image.channel, timestampNs);
    const depth = this.depth.get(key);
    if (depth) {
      this.depth.delete(key);
      this.onPair({ image, depth });
      return;
    }
    this.images.set(key, image);
    this._trim(this.images);
  }

  pushDepth(depth) {
    const timestampNs = frameTimestamp(depth);
    if (!depth || timestampNs <= 0n) return;
    const key = rgbdKey(depth.sourceChannel, timestampNs);
    const image = this.images.get(key);
    if (image) {
      this.images.delete(key);
      this.onPair({ image, depth });
      return;
    }
    this.depth.set(key, depth);
    this._trim(this.depth);
  }

  clear() {
    this.images.clear();
    this.depth.clear();
  }
}

export function isUint16MetricDepth(frame) {
  return Number(frame?.encoding) === DEPTH_ENCODING.UINT16_MILLIMETERS;
}
